domains/client/domain-operator/src/snap_sync.rs

```diff
@@ -0,0 +1,481 @@
+                                trace!(%block_number, "Domain block parsing result: {:?}", blocks);
+
+                                return blocks.pop().ok_or_else(|| {
+                                    sp_blockchain::Error::Application(
+                                        "Got empty state blocks collection for domain snap sync"
+                                            .into(),
+                                    )
+                                });
+                            }
+                            Err(error) => {
+                                error!(%block_number, ?error, "Domain block parsing error");
+                                continue;
+                            }
+                        }
+                    }
+                    Err(error) => {
+                        error!(%block_number, ?error, "Domain block sync error (inner)");
+                        continue;
+                    }
+                }
+            }
+            Err(error) => {
+                error!(%block_number, ?error, "Domain block sync error");
+                continue;
+            }
+        }
+    }
+
+    Err(sp_blockchain::Error::Application(
+        format!("Failed to get block {}", block_number).into(),
+    ))
+}
+
+fn convert_block_number<Block: BlockT>(block_number: NumberFor<Block>) -> u32 {
+    let block_number: u32 = match block_number.try_into() {
+        Ok(block_number) => block_number,
+        Err(_) => {
+            panic!("Can't convert block number.")
+        }
+    };
+
+    block_number
+}
+
+#[allow(clippy::too_many_arguments)]
+pub(crate) async fn snap_sync<Block, Client, NR, CBlock, CClient, Backend, CNR, AS>(
+    sync_params: SyncParams<Client, CClient, NR, Block, CBlock, CNR, AS>,
+) -> Result<(), sp_blockchain::Error>
+where
+    Block: BlockT,
+    Backend: sc_client_api::Backend<Block>,
+    Client: HeaderBackend<Block>
+        + BlockImport<Block>
+        + AuxStore
+        + ProofProvider<Block>
+        + ClientExt<Block, Backend>
+        + Send
+        + Sync
+        + 'static,
+    for<'a> &'a Client: BlockImport<Block>,
+    NR: NetworkRequest + Send + Sync,
+    CNR: NetworkRequest + Send + Sync,
+    CBlock: BlockT,
+    CClient: HeaderBackend<CBlock> + ProofProvider<CBlock> + Send + Sync + 'static,
+    AS: AuxStore,
+{
+    let execution_receipt_result = sync_params
+        .consensus_chain_sync_params
+        .execution_receipt_provider
+        .get_execution_receipt(None)
+        .await;
+    debug!(
+        "Snap-sync: execution receipt result - {:?}",
+        execution_receipt_result
+    );
+
+    let Some(last_confirmed_block_receipt) = execution_receipt_result else {
+        return Err(sp_blockchain::Error::RemoteFetchFailed);
+    };
+
+    if last_confirmed_block_receipt.domain_block_number == 0u32.into() {
+        return Err(sp_blockchain::Error::Application(
+            "Can't snap sync from genesis.".into(),
+        ));
+    }
```

Comment:
This can happen if the domain is instantiated but has produced less than 14_400 blocks, especially when we enable the domain in the mainnet phase 2, the operator node will have to use full sync and sync from genesis.

Is it possible to avoid this error? by either downloading the genesis state from other peers or deriving the genesis state from the consensus chain (after consensus sync is finished) as the domain bootstrapper currently does.